### ISSUE TITLE
Serialisation of miden assembly

### ISSUE THREAD

**Issue by frisitano** [maticnetwork/miden#225]
Hi,

I believe miden assembly is intended to be easily serialised as bytecode.  However when when exploring the code I can't see an implementation of assembly serialisation.  Has assembly serialisation been implemented?  If not I would be happy to take a go at and would appreciate if you could provide some pointers of how a solution should look.  I'm guessing we would would to implement the `IntoBytes` trait for a `Script`?

**Comment by bobbinth** [comment]
Hi! Serialization hasn't been implemented yet, but it's definitely something that would be good to implement.

However, it is still an open question about how to do it. The thing is that serializing the `Script` struct would not give good results as a lot of info from the source assembly file is not retained there. For example, all procedures are inlined, and thus, the resulting serialized version could be much bigger than the source version.

There probably should be something in-between the source text file and `Script` struct, but we haven't implemented this yet. The pipeline should look something like this:
```
[source .masm] -> [parsed AST] -> [Script]
```

We could probably start building `[source .masm] -> [parsed AST] ` step even now. There are a few considerations to keep in mind:
- The parser we implement should ideally be built in such a way that we could use it for something like a VS Code [LSP extension](https://code.visualstudio.com/api/language-extensions/language-server-extension-guide) in the future.
- Miden assembly has pretty simple syntax - so, we could do the whole parser ourselves - but I wonder if using something like [Pest](https://crates.io/crates/pest) or [rowan](https://crates.io/crates/rowan) would be a better idea.

**Comment by frisitano** [comment]
It makes sense to me that `[source .masm] -> [pased AST]` is lossless however I'm not clear on how the serialized AST would be used in practice.  Naively I would assume that if we wanted to submit a `create` extrinsic which stores a miden program in on-chain storage then we would want to do this with a serialized `Script`.  If we were to store serialized AST then we would have to compile the AST to a `Script` every time we retrieve the AST from on-chain storage and wanted to use it?

**Comment by bobbinth** [comment]
My original intent for Miden assembly was actually to have it stored on chain for public accounts.

The main reason is that in many cases it should be significantly more compact than the serialized version of the `Script` struct. For example, a call to a procedure from the standard library would serialize as 33 bytes. But an inlined procedure (e.g., [blake3](https://github.com/maticnetwork/miden/blob/next/stdlib/asm/crypto/hashes/blake3.masm) or [sha256](https://github.com/maticnetwork/miden/blob/next/stdlib/asm/crypto/hashes/sha256.masm)) may require many hundreds of bytes. Now, this would not always be true, small procedures such as [u64 arithmetic](https://github.com/maticnetwork/miden/blob/next/stdlib/asm/math/u64.masm) might actually be cheaper to inline. This could lead to some interesting optimizations, but that's something that can be considered later.

You are correct in that there would be some overhead in converting a serialized AST to a `Script` struct, but my thinking is that it shouldn't be too bad for the following reasons:
1. Converting AST into script should be relatively light-weight and could be done in parallel for many contracts.
2. For frequently-used procedures (and maybe even entire contracts), a node could do these conversions only once and then store the results in memory, thus further reducing compilation latency/overhead.

**Comment by frisitano** [comment]
This makes sense, thank you for explaining!  I'll take a take a look the libraries you suggested.  However I don't have any experience working with AST's so can't make a commitment to picking this up just yet.

**Comment by bobbinth** [comment]
@frisitano - no worries! I'll write up a separate issue specifically about a parser - and if you feel like you'd like to take a stab at it, just let me know.
